utility/SPIDEV/gpio.cpp

```diff
@@ -6,12 +6,63 @@
  *
  */
 #include <linux/gpio.h>
-#include "gpio.h"
 #include <unistd.h>    // close()
 #include <fcntl.h>     // open()
 #include <sys/ioctl.h> // ioctl()
+#include <errno.h>     // errno, strerror()
+#include <string.h>    // std::string, strcpy()
+#include <map>
+#include "gpio.h"
+
+// instantiate some global structs to setup cache
+// doing this globally ensures the data struct is zero-ed out
+typedef int gpio_fd; // for readability
+std::map<rf24_gpio_pin_t, gpio_fd> cached_pins;
+struct gpio_v2_line_request request;
+struct gpio_v2_line_values data;
+
+// use this struct to keep a track of open file descriptors
+struct GlobalCache
+{
+    const char* chip = "/dev/gpiochip4";
+    int fd = -1;
 
-const char* dev_name = "/dev/gpiochip4";
+    // should be called automatically on program start.
+    // Here, we do some one-off configuration.
+    GlobalCache()
+    {
+        if (fd < 0) {
+            fd = ::open(chip, O_RDWR);
+            if (fd < 0) {
+                chip = "/dev/gpiochip0";
+                fd = ::open(chip, O_RDWR);
+                if (fd < 0) {
+                    std::string msg = "Can't open device ";
+                    msg += chip;
+                    throw GPIOException(msg);
+                }
+            }
+        }
```

Comment:
We should find a way that allows users to specify a different `/dev/gpiochipX` for devices that don't have/use `gpiochip4` or `gpiochip0`. I don't have any hardware that I could test this on, but I imagine nVidia's Jetson family would be a prime candidate because it has multiple GPIO chips exposing multiple GPIO buses.

I'm thinking a new build option is in order. Something like:
```
cmake ../RF24 -DRF24_DRIVER=SPIDEV -DRF24_SPIDEV_GPIO_CHIP="/dev/gpiochip3"
```

Reply:
Too bad we can't automate it more, it would be nice to just autodetect it, but this seems like a viable option.

Reply:
I think there is a way to check if the pin number exists on a certain GPIO chip using the `ioctl()` (I think it's defined as `GPIO_V2_GET_DEVICE_INFO`). But at a certain point, the user still needs to know a bit about the different GPIO chips available for their board.

Reply:
I added `RF24_SPIDEV_GPIO_CHIP` define. This can only be configured when compiling the library (not on individual app executables).

This is undocumented for now. It should be possible to configure it using the older ./configure script if the make command is passed
```shall
make -DRF24_SPIDEV_GPIO_CHIP="/dev/gpiochipX"
```
but I haven't tested that. I have tested it using
```
cd RF24/build
cmake .. -DRF24_DRIVER=SPIDEV -DRF24_SPIDEV_GPIO_CHIP="/dev/gpiochip0"
make VERBOSE=1
``` 
when the fallback in code was set to `/dev/gpiochip1`.

---
> I think there is a way to check if the pin number exists on a certain GPIO chip

I also added a check to make sure the pin number is available for the given GPIO chip:
https://github.com/nRF24/RF24/blob/4e82d88e4f7ed78522cf400d1e0386f4d6e681cc/utility/SPIDEV/gpio.cpp#L23
in GPIOCache c'tor:
https://github.com/nRF24/RF24/blob/4e82d88e4f7ed78522cf400d1e0386f4d6e681cc/utility/SPIDEV/gpio.cpp#L70-L77
and
https://github.com/nRF24/RF24/blob/4e82d88e4f7ed78522cf400d1e0386f4d6e681cc/utility/SPIDEV/gpio.cpp#L103-L109

Reply:
Just so I understand, will the old install script still work without the define? Will it fall back to using chip0 or 4? I think that is the preferred behavior so we don't break the old script.

Reply:
Yeah, it will use gpiochip4 if not defined and fallback to gpiochip0 for backward compatibility. This is true despite the build system generator used.
https://github.com/nRF24/RF24/blob/4e82d88e4f7ed78522cf400d1e0386f4d6e681cc/utility/SPIDEV/gpio.h#L24-L30
https://github.com/nRF24/RF24/blob/4e82d88e4f7ed78522cf400d1e0386f4d6e681cc/utility/SPIDEV/gpio.cpp#L28